node/core/approval-voting/src/lib.rs

```diff
@@ -1192,12 +1169,66 @@ async fn handle_from_overseer<Context>(
 
 				Vec::new()
 			},
+			ApprovalVotingMessage::GetApprovalSignaturesForCandidate(candidate_hash, tx) => {
+				let votes = get_approval_signatures_for_candidate(ctx, db, candidate_hash).await?;
+				if let Err(_) = tx.send(votes) {
+					gum::debug!(
+						target: LOG_TARGET,
+						"Sending approval signatures back failed, as receiver got closed"
+					);
+				}
+				Vec::new()
+			},
 		},
 	};
 
 	Ok(actions)
 }
 
+/// Retrieve approval signatures.
+///
+/// This involves an unbounded message send to approval-distribution, the caller has to ensure that
+/// calls to this function are infrequent and bounded.
+#[overseer::contextbounds(ApprovalVoting, prefix = self::overseer)]
+async fn get_approval_signatures_for_candidate<Context>(
+	ctx: &mut Context,
+	db: &OverlayedBackend<'_, impl Backend>,
+	candidate_hash: CandidateHash,
+) -> SubsystemResult<HashMap<ValidatorIndex, ValidatorSignature>> {
+	let entry = match db.load_candidate_entry(&candidate_hash)? {
+		None => return Ok(HashMap::new()),
+		Some(e) => e,
+	};
+
+	let relay_hashes = entry.block_assignments.iter().map(|(relay_hash, _)| relay_hash);
+
+	let mut candidate_indices = HashSet::new();
+	// Retrieve `CoreIndices`/`CandidateIndices` as required by approval-distribution:
+	for hash in relay_hashes {
+		let entry = match db.load_block_entry(hash)? {
+			None => continue,
+			Some(e) => e,
+		};
+		for (candidate_index, (_core_index, c_hash)) in entry.candidates().iter().enumerate() {
+			if c_hash == &candidate_hash {
+				candidate_indices.insert((*hash, candidate_index as u32));
+				break
+			}
+		}
+	}
+
+	let (tx, rx) = oneshot::channel();
+	// We should not be sending this message frequently - caller must make sure this is bounded.
```

Comment:
What does frequently mean in this context ?  Also does it make sense to time this call ?

Reply:
Should we have a threshold and metric to reflect it being breached ?

Reply:
Good idea! will add!